CONTRIBUTING.md

```diff
@@ -0,0 +1,900 @@
+
+Avoid semicolons. (A few resources to help you understand this: [*](http://blog.izs.me/post/2353458699/an-open-letter-to-javascript-leaders-regarding) [*](http://inimino.org/~inimino/blog/javascript_semicolons) [*](http://jamesallardice.com/understanding-automatic-semi-colon-insertion-in-javascript/))
+
+### Redundant expressions
+
+Don't use redundant parentheses, curly braces, `undefined`s, labels, ternaries, etc. Examples:
+
+- `var foo = value === 1 ? true : false` is better written as `var foo = value === 1`.
+- `var foo = (value === 1) ? bar : baz` is better written as `var foo = value === 1 ? bar : baz`.
+- `var foo = undefined` is better written as `var foo`.
+- `return undefined` is better written as `return`.
+- `delete(obj.foo)` is better written as `delete obj.foo`.
+- `return` at the end of a function if you're returning `undefined` should just be omitted.
+- If you're returning `null`, there's usually little reason to not just return `undefined`.
+- Unused variables should be removed.
+- An exception is detailed below for assignment in conditions.
+
+### Equality testing
+
+Use strict equality (`===`/`!==`), not loose equality (`==`/`!=`). The only exception is if you're testing for `null` or `undefined`, in which loose equality, `value == null` and `value != null` is preferred. The reason is for type safety with primitives.
+
+### `eval` and friends
+
+Never use `eval`, the `Function` constructor, and friends. Those will fail when running under CSP restrictions, and this library should work in those environments.
+
+### Quotes
+
+Prefer double quotes to single quotes, but using single quotes to avoid escaping is okay.
+
+### Strict mode
+
+All code must be in strict mode, preferably wrapped in an IIFE.
+
+## Comments
+
+Comments are helpful. Use descriptive comments where needed, so others can read it and understand it. If a non-obvious hack is used, explain it with a comment. But don't repeat yourself with a redundant comment when the code adequately describes itself.
+
+```js
+// Good, code is self-descriptive
+var CARD_DECK_SIZE = 52
+
+// Shuffle the deck of cards.
+for (var i = 0; i < CARD_DECK_SIZE; i++) {
+	var j = i + (Math.random() * CARD_DECK_SIZE - i)|0
+
+	var tmp = deck[i]
+	deck[i] = deck[j]
+	deck[j] = tmp
+}
+
+// Also good, requires a descriptive comment.
+function toFastProperties(obj) {
+	// Bluebird's toFastProperties hack. Forces V8 to optimize object as prototype,
+	// significantly speeding up property access.
+
+	/* eslint-disable no-eval */
+	function C() {}
+	C.prototype = obj
+	new C()
+	return
+	eval(obj)
+	/* eslint-enable no-eval */
+}
+
+// Bad, redundant comments
+var CARD_DECK_SIZE = 52
+
+// Shuffle the deck of cards.
+function shuffle(deck) {
+	for (var i = 0; i < CARD_DECK_SIZE; i++) {
+		// Generate a random card index to swap at.
+		var j = i + (Math.random() * CARD_DECK_SIZE - i)|0
+
+		// Swap the cards at each index.
+		var tmp = deck[i]
+		deck[i] = deck[j]
+		deck[j] = tmp
+	}
+}
+```
```

Comment:
pull comments about comments out of comments (i.e. `Good, code is self-descriptive` is not a good comment itself).